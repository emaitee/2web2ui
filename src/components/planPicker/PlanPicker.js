import React, { Component } from 'react';
import Downshift from 'downshift';
import cx from 'classnames';

import { Icon } from '@sparkpost/matchbox';
import Plan from './Plan';
import styles from './PlanPicker.module.scss';

class PlanPicker extends Component {
  handleOpen = () => {
    this.input.focus();
  }

  planFn = ({
    getInputProps,
    getRootProps,
    getButtonProps,
    getItemProps,
    isOpen,
    selectedItem,
    highlightedIndex
  }) => {
    const { plans, input } = this.props;

    const items = plans.map((item, index) => {
      const classes = cx(
        styles.DropdownPlan,
        selectedItem.code === item.code && styles.selected,
        highlightedIndex === index && styles.highlighted
      );

      return <Plan key={index} className={classes} {...getItemProps({ item, index, plan: item })} />;
    });

    const listClasses = cx(styles.List, isOpen && styles.open);
    const triggerClasses = cx(styles.TriggerPlan, isOpen && styles.triggerOpen);
    const triggerProps = getButtonProps({
      plan: selectedItem,
      onClick: this.handleOpen
    });

    return (
       <div {...getRootProps()} className={styles.PlanPicker}>
         <div className={listClasses}>{ items }</div>
         <Icon name='ChevronDown' size={24} className={styles.Chevron} />
         <input {...getInputProps()} ref={(input) => this.input = input} className={styles.Input} />
         <Plan {...triggerProps} className={triggerClasses} />
       </div>
    );
  };

  render() {
    const { plans, input } = this.props;
    const { onChange, value } = input;

    return (
      <Downshift
        onChange={onChange}
        defaultSelectedItem={value || plans[0]} >
        {this.planFn}
      </Downshift>
    );
  }
}

export default PlanPicker;
