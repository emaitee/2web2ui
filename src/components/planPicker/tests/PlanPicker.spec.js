import { mount } from 'enzyme';
import React from 'react';
import Downshift from 'downshift';

import PlanPicker from '../PlanPicker';

const plans = [
  {
    code: '1',
    includesIp: true,
    monthly: 100,
    name: 'One',
    overage: 0.1,
    volume: 1
  },
  {
    code: '2',
    includesIp: false,
    monthly: 0,
    name: 'Two',
    overage: 0.2,
    volume: 2,
    isFree: true
  },
  {
    code: '3',
    monthly: 300,
    name: 'Three',
    overage: 0.3,
    volume: 3,
  },
];

const selected = {
  code: '4',
  monthly: 400,
  name: 'Four',
  overage: 0.4,
  volume: 4,
};

const props = {
  input: { onChange: jest.fn() },
  plans,
};

const selectedProps = { ...props, input: { ...props.input, value: selected } };

it('renders correctly', () => {
  const wrapper = mount(<PlanPicker {...props} />);
  expect(wrapper).toMatchSnapshot();
});

it('renders correctly with an initial value', () => {
  const wrapper = mount(<PlanPicker {...selectedProps} />);
  expect(wrapper).toMatchSnapshot();
});

it('should focus on input when opened', () => {
  const wrapper = mount(<PlanPicker {...props} />);
  const openSpy = jest.spyOn(wrapper.instance(), 'handleOpen');
  expect(openSpy).not.toHaveBeenCalled();

  // This passes and I have no idea why it works
  wrapper.find(Downshift).node.openMenu();
  wrapper.find('a').at(3).simulate('click');
  expect(openSpy).toHaveBeenCalled();
});
