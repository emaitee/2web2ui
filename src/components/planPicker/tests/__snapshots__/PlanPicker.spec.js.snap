// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<PlanPicker
  input={
    Object {
      "onChange": [Function],
    }
  }
  plans={
    Array [
      Object {
        "code": "1",
        "includesIp": true,
        "monthly": 100,
        "name": "One",
        "overage": 0.1,
        "volume": 1,
      },
      Object {
        "code": "2",
        "includesIp": false,
        "isFree": true,
        "monthly": 0,
        "name": "Two",
        "overage": 0.2,
        "volume": 2,
      },
      Object {
        "code": "3",
        "monthly": 300,
        "name": "Three",
        "overage": 0.3,
        "volume": 3,
      },
    ]
  }
>
  <Downshift
    defaultHighlightedIndex={null}
    defaultInputValue=""
    defaultIsOpen={false}
    defaultSelectedItem={
      Object {
        "code": "1",
        "includesIp": true,
        "monthly": 100,
        "name": "One",
        "overage": 0.1,
        "volume": 1,
      }
    }
    getA11yStatusMessage={[Function]}
    id="downshift-1"
    itemToString={[Function]}
    onChange={[Function]}
    onStateChange={[Function]}
    onUserAction={[Function]}
  >
    <div>
      <div
        className=""
      >
        <Plan
          className=""
          id="downshift-1-item-0"
          onClick={[Function]}
          onMouseEnter={[Function]}
          plan={
            Object {
              "code": "1",
              "includesIp": true,
              "monthly": 100,
              "name": "One",
              "overage": 0.1,
              "volume": 1,
            }
          }
        >
          <a
            className=""
            id="downshift-1-item-0"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            <span>
              <strong>
                1
              </strong>
               emails for 
              <span>
                <strong>
                  $
                  100
                </strong>
                /mo
              </span>
            </span>
            <span>
              $0.10/ thousand extra emails
              , first dedicated IP address is free
            </span>
          </a>
        </Plan>
        <Plan
          className=""
          id="downshift-1-item-1"
          onClick={[Function]}
          onMouseEnter={[Function]}
          plan={
            Object {
              "code": "2",
              "includesIp": false,
              "isFree": true,
              "monthly": 0,
              "name": "Two",
              "overage": 0.2,
              "volume": 2,
            }
          }
        >
          <a
            className=""
            id="downshift-1-item-1"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            <span>
              <strong>
                2
              </strong>
               emails for 
              <strong>
                Free
              </strong>
            </span>
            <span>
              Full-featured developer account
            </span>
          </a>
        </Plan>
        <Plan
          className=""
          id="downshift-1-item-2"
          onClick={[Function]}
          onMouseEnter={[Function]}
          plan={
            Object {
              "code": "3",
              "monthly": 300,
              "name": "Three",
              "overage": 0.3,
              "volume": 3,
            }
          }
        >
          <a
            className=""
            id="downshift-1-item-2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            <span>
              <strong>
                3
              </strong>
               emails for 
              <span>
                <strong>
                  $
                  300
                </strong>
                /mo
              </span>
            </span>
            <span>
              $0.30/ thousand extra emails
            </span>
          </a>
        </Plan>
      </div>
      <t
        name="ChevronDown"
        size={24}
      >
        <MdExpandMore
          size={24}
        >
          <IconBase
            size={24}
            viewBox="0 0 40 40"
          >
            <svg
              fill="currentColor"
              height={24}
              preserveAspectRatio="xMidYMid meet"
              style={
                Object {
                  "color": undefined,
                  "verticalAlign": "middle",
                }
              }
              viewBox="0 0 40 40"
              width={24}
            >
              <g>
                <path
                  d="m27.7 14.3l2.3 2.3-10 10-10-10 2.3-2.3 7.7 7.7z"
                />
              </g>
            </svg>
          </IconBase>
        </MdExpandMore>
      </t>
      <input
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-expanded={false}
        autoComplete="off"
        id="downshift-input-2"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        role="combobox"
        value="[object Object]"
      />
      <Plan
        aria-expanded={false}
        aria-haspopup={true}
        aria-label="open menu"
        className=""
        onClick={[Function]}
        onKeyDown={[Function]}
        plan={
          Object {
            "code": "1",
            "includesIp": true,
            "monthly": 100,
            "name": "One",
            "overage": 0.1,
            "volume": 1,
          }
        }
        role="button"
      >
        <a
          aria-expanded={false}
          aria-haspopup={true}
          aria-label="open menu"
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          <span>
            <strong>
              1
            </strong>
             emails for 
            <span>
              <strong>
                $
                100
              </strong>
              /mo
            </span>
          </span>
          <span>
            $0.10/ thousand extra emails
            , first dedicated IP address is free
          </span>
        </a>
      </Plan>
    </div>
  </Downshift>
</PlanPicker>
`;

exports[`renders correctly with an initial value 1`] = `
<PlanPicker
  input={
    Object {
      "onChange": [Function],
      "value": Object {
        "code": "4",
        "monthly": 400,
        "name": "Four",
        "overage": 0.4,
        "volume": 4,
      },
    }
  }
  plans={
    Array [
      Object {
        "code": "1",
        "includesIp": true,
        "monthly": 100,
        "name": "One",
        "overage": 0.1,
        "volume": 1,
      },
      Object {
        "code": "2",
        "includesIp": false,
        "isFree": true,
        "monthly": 0,
        "name": "Two",
        "overage": 0.2,
        "volume": 2,
      },
      Object {
        "code": "3",
        "monthly": 300,
        "name": "Three",
        "overage": 0.3,
        "volume": 3,
      },
    ]
  }
>
  <Downshift
    defaultHighlightedIndex={null}
    defaultInputValue=""
    defaultIsOpen={false}
    defaultSelectedItem={
      Object {
        "code": "4",
        "monthly": 400,
        "name": "Four",
        "overage": 0.4,
        "volume": 4,
      }
    }
    getA11yStatusMessage={[Function]}
    id="downshift-1"
    itemToString={[Function]}
    onChange={[Function]}
    onStateChange={[Function]}
    onUserAction={[Function]}
  >
    <div>
      <div
        className=""
      >
        <Plan
          className=""
          id="downshift-1-item-0"
          onClick={[Function]}
          onMouseEnter={[Function]}
          plan={
            Object {
              "code": "1",
              "includesIp": true,
              "monthly": 100,
              "name": "One",
              "overage": 0.1,
              "volume": 1,
            }
          }
        >
          <a
            className=""
            id="downshift-1-item-0"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            <span>
              <strong>
                1
              </strong>
               emails for 
              <span>
                <strong>
                  $
                  100
                </strong>
                /mo
              </span>
            </span>
            <span>
              $0.10/ thousand extra emails
              , first dedicated IP address is free
            </span>
          </a>
        </Plan>
        <Plan
          className=""
          id="downshift-1-item-1"
          onClick={[Function]}
          onMouseEnter={[Function]}
          plan={
            Object {
              "code": "2",
              "includesIp": false,
              "isFree": true,
              "monthly": 0,
              "name": "Two",
              "overage": 0.2,
              "volume": 2,
            }
          }
        >
          <a
            className=""
            id="downshift-1-item-1"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            <span>
              <strong>
                2
              </strong>
               emails for 
              <strong>
                Free
              </strong>
            </span>
            <span>
              Full-featured developer account
            </span>
          </a>
        </Plan>
        <Plan
          className=""
          id="downshift-1-item-2"
          onClick={[Function]}
          onMouseEnter={[Function]}
          plan={
            Object {
              "code": "3",
              "monthly": 300,
              "name": "Three",
              "overage": 0.3,
              "volume": 3,
            }
          }
        >
          <a
            className=""
            id="downshift-1-item-2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            <span>
              <strong>
                3
              </strong>
               emails for 
              <span>
                <strong>
                  $
                  300
                </strong>
                /mo
              </span>
            </span>
            <span>
              $0.30/ thousand extra emails
            </span>
          </a>
        </Plan>
      </div>
      <t
        name="ChevronDown"
        size={24}
      >
        <MdExpandMore
          size={24}
        >
          <IconBase
            size={24}
            viewBox="0 0 40 40"
          >
            <svg
              fill="currentColor"
              height={24}
              preserveAspectRatio="xMidYMid meet"
              style={
                Object {
                  "color": undefined,
                  "verticalAlign": "middle",
                }
              }
              viewBox="0 0 40 40"
              width={24}
            >
              <g>
                <path
                  d="m27.7 14.3l2.3 2.3-10 10-10-10 2.3-2.3 7.7 7.7z"
                />
              </g>
            </svg>
          </IconBase>
        </MdExpandMore>
      </t>
      <input
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-expanded={false}
        autoComplete="off"
        id="downshift-input-3"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        role="combobox"
        value="[object Object]"
      />
      <Plan
        aria-expanded={false}
        aria-haspopup={true}
        aria-label="open menu"
        className=""
        onClick={[Function]}
        onKeyDown={[Function]}
        plan={
          Object {
            "code": "4",
            "monthly": 400,
            "name": "Four",
            "overage": 0.4,
            "volume": 4,
          }
        }
        role="button"
      >
        <a
          aria-expanded={false}
          aria-haspopup={true}
          aria-label="open menu"
          className=""
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          <span>
            <strong>
              4
            </strong>
             emails for 
            <span>
              <strong>
                $
                400
              </strong>
              /mo
            </span>
          </span>
          <span>
            $0.40/ thousand extra emails
          </span>
        </a>
      </Plan>
    </div>
  </Downshift>
</PlanPicker>
`;
